src/gpu/intel/jit/ir/include/object.hpp

```diff
@@ -0,0 +1,365 @@
+    //       the reference dangling due to the destruction of the former object;
+    //       please only call this method on non-const objects if absolutely
+    //       necessary, and please don't add a non-const variant of the method!
+    template <typename T>
+    const T &as() const {
+        gpu_assert(is<T>());
+        return *as_ptr<T>(); // fails on incorrect casts even in Release
+    }
+
+    // Downcasts the object to the IR type, returns a pointer. If the IR type
+    // doesn't match the real IR type, returns nullptr.
+    // N.B.: this can potentially be dangerous if applied to non-const objects,
+    //       since assigning a different value to the source object might make
+    //       the reference dangling due to the destruction of the former object;
+    //       please only call this method on non-const objects if absolutely
+    //       necessary, and please don't add a non-const variant of the method!
+    template <typename T>
+    const T *as_ptr() const {
+        return (is<T>()) ? (const T *)this : nullptr;
+    }
+
+    // Returns true if T matches the real IR type.
+    template <typename T>
+    bool is() const {
+        return info() == T::get_info();
+    }
+
+    virtual std::string str() const;
+
+    virtual object_t _mutate(ir_mutator_t &mutator) const;
+    virtual void _visit(ir_visitor_t &visitor) const;
+
+protected:
+    friend class dnnl::impl::gpu::intel::jit::object_t;
+    template <typename T>
+    friend struct info_t;
+
+    void retain() { ref_count_.increment(); }
+    void release() {
+        if (ref_count_.decrement() == 0) { delete this; }
+    }
+
+    struct info_t {
+        constexpr info_t(void *uid, bool is_expr, bool is_stmt)
+            : uid(uid), is_expr(is_expr), is_stmt(is_stmt) {}
+
+        void *uid = nullptr;
+        bool is_expr = false;
+        bool is_stmt = false;
+
+        bool operator==(const info_t &other) const { return uid == other.uid; }
+        bool operator!=(const info_t &other) const {
+            return !operator==(other);
+        }
+    };
+
+    impl_t(info_t info) : info_(info) {};
+
+private:
+    // Reference counter for IR objects.
+    class ref_count_t {
+    public:
+        ref_count_t() : value_(0) {}
+        ref_count_t(const ref_count_t &) = delete;
+        ref_count_t &operator=(const ref_count_t &) = delete;
+        ~ref_count_t() = default;
+
+        uint32_t increment() { return ++value_; }
+        uint32_t decrement() { return --value_; }
+
+    private:
+        uint32_t value_;
+    };
+
+    // Type information.
+    const info_t &info() const { return info_; };
+
+    ref_count_t ref_count_;
+    info_t info_;
+};
+
+template <typename T>
+struct info_t {
+    using self_type = T;
+
+protected:
+    friend class impl_t;
+    static impl_t::info_t get_info() {
+        return impl_t::info_t(
+                (void *)get_info, is_expr_t<T>::value, is_stmt_t<T>::value);
```

Comment:
I remember something like that was used in the first version but it caused issues due to MSVC folding multiple functions into one: https://github.com/uxlfoundation/oneDNN/commit/6383a4f1ad

I tried a toy example with MSVC - observing this folding even with default options: https://godbolt.org/z/44eaPWonG

Please take a look if it's an issue here. Looks like the trick with a static variable like in the commit could help.

Reply:
Thanks for pointing this out, I had thought there was likely to be some issue with here investigate further.  I went ahead and implemented a registry system which should provide effectively the same performance while avoiding all the pitfalls of function pointers.